modules/glfw/src/filewatcher.cpp

```diff
@@ -0,0 +1,325 @@
+        auto range = util::zip(handles_, observed_);
+        auto it = std::remove_if(range.begin(), range.end(), [&](auto&& elem) {
+            return std::find(toRemove.begin(), toRemove.end(), elem.second().first) !=
+                   toRemove.end();
+        });
+        for (auto&& [handle, observed] : util::as_range(it, range.end())) {
+            FindCloseChangeNotification(handle);
+            observed.first.clear();
+            observed.second.clear();
+            --active_;
+        }
+    }
+
+    void add(const std::vector<std::string>& toAdd) {
+        for (auto& path : toAdd) {
+            auto wpath = util::toWstring(path);
+            const auto handle = FindFirstChangeNotification(wpath.c_str(), TRUE, filter);
+
+            if (handle == INVALID_HANDLE_VALUE || handle == nullptr) {
+                LogError("FindFirstChangeNotification function failed.");
+                continue;
+            }
+
+            handles_[active_] = handle;
+            observed_[active_].first = path;
+            for (auto& elem : filesystem::getDirectoryContentsRecursively(
+                     path, filesystem::ListMode::FilesAndDirectories)) {
+                observed_[active_].second[elem] = filesystem::fileModificationTime(elem);
+            }
+
+            ++active_;
+        }
+    }
+
+    void watch() {
+        while (!stop_) {
+            {
+                std::scoped_lock lock{mutex_};
+                if (!toRemove_.empty()) {
+                    remove(toRemove_);
+                    toRemove_.clear();
+                }
+                if (!toAdd_.empty()) {
+                    add(toAdd_);
+                    toAdd_.clear();
+                }
+            }
+            if (active_ == 0) {
+                std::this_thread::sleep_for(timeout_);
+            } else {
+                const auto status =
+                    WaitForMultipleObjects(static_cast<DWORD>(active_), handles_.data(), FALSE,
+                                           static_cast<DWORD>(timeout_.count()));
+                if (status >= WAIT_OBJECT_0 && status < WAIT_OBJECT_0 + MAXIMUM_WAIT_OBJECTS) {
+
+                    const auto& path = observed_[status - WAIT_OBJECT_0].first;
+                    auto& files = observed_[status - WAIT_OBJECT_0].second;
+
+                    const auto changedFiles = getChangedAndUpdateFiles(path, files);
+                    for (auto&& [changedFile, action] : changedFiles) {
+                        changeCallback_(path, changedFile, action);
+                    }
+
+                    const auto handle = handles_[status - WAIT_OBJECT_0];
+                    FindNextChangeNotification(handle);
+                }
+            }
+        }
+        for (auto handle : util::as_range(handles_.begin(), handles_.begin() + active_)) {
+            FindCloseChangeNotification(handle);
+        }
+    }
+
+    std::vector<std::pair<std::string, Action>> getChangedAndUpdateFiles(
```

Comment:
`getChangedAndUpdatedFiles`  with "update_d_"?

Reply:
no it return the list of new changes, and updates the time stamps of the observed files